### PR TITLE
ENH: Add annotations to `scipy.spatial.geometric_slerp`

### PR DESCRIPTION
**Pull request by BvB93**
#### Reference issue
n.a.

#### What does this implement/fix?
This PR adds annotations to `scipy.spatial.geometric_slerp` for the purpose of static type checking.

#### Additional information
n.a.

### CODE REVIEW

File: scipy/spatial/_geometric_slerp.py
Comment (BvB93) [inline scipy/spatial/_geometric_slerp.py:39]: Note that an explicit `isinstance(t, float)` check is performed by the function, 
so we can't use [`DecimalNumber`](https://github.com/scipy/scipy/blob/2c72ae27e59e64f47d64bf65800a291ff972df86/scipy/_lib/_util.py#L20) here no matter if we'd want to.

File: scipy/spatial/_geometric_slerp.py
Comment (BvB93) [inline scipy/spatial/_geometric_slerp.py:40]: This can technically also be a list of arrays for as long as https://github.com/scipy/scipy/pull/14173 isn't merged.

Comment (BvB93) [reply scipy/spatial/_geometric_slerp.py:40]: Considering the mypy error: on second thought, let's wait until aforementioned PR is merged.

**Comment by BvB93** [comment]
With the merging of https://github.com/scipy/scipy/pull/14173 this PR can be considered ready for review.

**Review by tylerjereddy — APPROVED** [review]
LGTM now, thanks Bas

**Pull request merged**
